### PR TITLE
Add Google Fonts provider and register a selection of fonts

### PR DESCRIPTION
**Pull request by creativecoder**
#### Changes proposed in this Pull Request:

* Adds a new Jetpack package with a font provider for Google fonts, based on the work from https://github.com/WordPress/wordpress-develop/pull/1736, and using the Simplified API from https://github.com/WordPress/gutenberg/pull/37140.
* Adds a Jetpack plugin module that registers a selection of Google Fonts from the Jetpack Global Styles feature (https://github.com/Automattic/wp-calypso/blob/3fff686fba8d74690cd03ec5f10ac43ec3e046b3/apps/editing-toolkit/editing-toolkit-plugin/global-styles/class-global-styles.php#L84)

#### Jetpack product discussion

paYE8P-1qo-p2

#### Does this pull request change what data or activity we track or use?

Loads Google Fonts settings and font files using Google APIs, if the module is activated.

#### Testing instructions:

Install the Gutenberg plugin with changes from this PR https://github.com/WordPress/gutenberg/pull/37140 (now in Gutenberg `trunk`, but not yet released).

* `pnpm jetpack install -r && pnpm jetpack install --all` to install the new fonts provider package.
* Activate the Jetpack plugin and activate the Google Fonts module from `/wp-admin/admin.php?page=jetpack_modules`
* Activate a block based theme, such as Twenty Twenty Two
* Go to the Site editor and open the Global Styles > Text panel and select one of the new font choices.

image

* Save your changes and check that the font loads on the front end of the Site.
* Open the Post editor and add a block that has font styles, like a button block.
* Select one of the font options, save, and check the front-end.

image

### CODE REVIEW

File: projects/packages/google-fonts-provider/src/class-google-fonts-provider.php [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+/**
+ * WordPress webfonts provider for Google Fonts
+ *
+ * @package automattic/google-fonts-provider
+ * @since $$next-version$$
+ */
+
+namespace Automattic\Jetpack\Fonts;
+
+if ( ! class_exists( '\WP_Webfonts_Provider' ) ) {
```
Comment (zaguiini) [inline projects/packages/google-fonts-provider/src/class-google-fonts-provider.php:11]: If the class doesn't exist, we return early?

**Comment by creativecoder** [comment]
wpcom diff: D75501-code

File: projects/packages/google-fonts-provider/src/class-google-fonts-provider.php (hunk 1 above)
Comment (creativecoder) [reply projects/packages/google-fonts-provider/src/class-google-fonts-provider.php:11]: This is simply a way to stop loading the code in this file if the `WP_Webfonts_Provider` class isn't available (from Gutenberg/WP).

File: projects/packages/google-fonts-provider/composer.json [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+			"type": "path",
+			"url": "../../packages/*",
+			"options": {
+				"monorepo": true
+			}
+		}
+	],
+	"minimum-stability": "dev",
+	"prefer-stable": true,
+	"extra": {
```
Comment (jeherve) [inline projects/packages/google-fonts-provider/composer.json:38]:
```suggestion
	"extra": {
		"autotagger": true,
		"mirror-repo": "Automattic/jetpack-google-fonts-provider",
		"textdomain": "jetpack-google-fonts-provider",
		"changelogger": {
			"link-template": "https://github.com/Automattic/jetpack-google-fonts-provider/compare/v${old}...v${new}"
		},
```

File: projects/packages/google-fonts-provider/.phpcs.dir.xml [hunk 3]
```diff
@@ -0,0 +1,24 @@
+<?xml version="1.0"?>
+<ruleset>
+
+	<rule ref="WordPress.WP.I18n">
+		<properties>
+			<property name="text_domain" type="array">
+				<element value="package-name" />
```
Comment (jeherve) [inline projects/packages/google-fonts-provider/.phpcs.dir.xml:7]:
```suggestion
				<element value="jetpack-google-fonts-provider" />
```

File: projects/packages/google-fonts-provider/.phpcs.dir.xml [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+			</property>
+		</properties>
+	</rule>
+	<rule ref="Jetpack.Functions.I18n">
+		<properties>
+			<property name="text_domain" value="package-name" />
```
Comment (jeherve) [inline projects/packages/google-fonts-provider/.phpcs.dir.xml:13]:
```suggestion
			<property name="text_domain" value="jetpack-google-fonts-provider" />
```

File: projects/packages/google-fonts-provider/.phpcs.dir.xml [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+		</properties>
+	</rule>
+
+	<rule ref="WordPress.Utils.I18nTextDomainFixer">
+		<properties>
+			<property name="old_text_domain" type="array" />
+			<property name="new_text_domain" value="package-name" />
```
Comment (jeherve) [inline projects/packages/google-fonts-provider/.phpcs.dir.xml:20]:
```suggestion
			<property name="new_text_domain" value="jetpack-google-fonts-provider" />
```

File: projects/packages/google-fonts-provider/src/class-google-fonts-provider.php [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+	 *    4. Generates the `@font-face` for all its webfonts.
+	 *
+	 * @return string The `@font-face` CSS.
+	 */
+	public function get_css() {
+		$css  = '';
+		$urls = $this->build_collection_api_urls();
+
+		foreach ( $urls as $url ) {
+			$css .= $this->get_cached_remote_styles( 'google_fonts_' . md5( $url ), $url );
```
Comment (jeherve) [inline projects/packages/google-fonts-provider/src/class-google-fonts-provider.php:105]: Should we add a `jetpack_` prefix here, so site owners can find them easily, and so we can clean after ourselves easier?

File: projects/packages/google-fonts-provider/README.md [hunk 7]
```diff
@@ -0,0 +1,20 @@
+# google-fonts-provider
+
+WordPress Webfonts provider for Google Fonts
```
Comment (jeherve) [inline projects/packages/google-fonts-provider/README.md:3]: Maybe we can expand on this a bit, and mention the work in progress in the Gutenberg repo, and the previous P2 posts like [this one](https://make.wordpress.org/core/2021/09/28/implementing-a-webfonts-api-in-wordpress-core/)?

File: projects/plugins/jetpack/modules/google-fonts.php [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+/**
+ * Register a curated selection of Google Fonts.
+ *
+ * @return void
+ */
+function jetpack_add_google_font_provider() {
+	wp_register_webfont_provider( 'google', '\Automattic\Jetpack\Fonts\Google_Fonts_Provider' );
+
+	foreach ( JETPACK_GOOGLE_FONTS_LIST as $font_family ) {
```
Comment (jeherve) [inline projects/plugins/jetpack/modules/google-fonts.php:61]: We may have to introduce a filter for `JETPACK_GOOGLE_FONTS_LIST` quite early on; I would expect site owners to want to get rid of some specific fonts on their site to reduce clutter.

File: projects/plugins/jetpack/modules/google-fonts.php [hunk 9]
```diff
@@ -0,0 +1,72 @@
+<?php
+/**
+ * Module Name: Google Fonts
+ * Module Description: A selection of Google fonts for Block enabled themes.
+ * Sort Order: 1
+ * Recommendation Order: 2
+ * First Introduced: $$next-version$$
+ * Requires Connection: No
+ * Auto Activate: Yes
```
Comment (jeherve) [inline projects/plugins/jetpack/modules/google-fonts.php:9]: I don't know that I would auto-activate the module until we have a UI for it in the Jetpack > Dashboard, so folks can disable the feature if they'd like it.

**Review by jeherve — CHANGES_REQUESTED** [review]
Just made a few general comments. I think we could merge this early on and iterate later on, if you'd like to test this further.

File: projects/packages/google-fonts-provider/composer.json (hunk 2 above)
Comment (creativecoder) [reply projects/packages/google-fonts-provider/composer.json:38]: Done in https://github.com/Automattic/jetpack/pull/23045/commits/21e31d128ab3389d819a7a1314ee9d116f9fee61

File: projects/packages/google-fonts-provider/src/class-google-fonts-provider.php (hunk 6 above)
Comment (creativecoder) [reply projects/packages/google-fonts-provider/src/class-google-fonts-provider.php]: Good idea!

File: projects/packages/google-fonts-provider/README.md (hunk 7 above)
Comment (creativecoder) [reply projects/packages/google-fonts-provider/README.md:3]: https://github.com/Automattic/jetpack/pull/23045/commits/9341447ae50d280364b3e8f6e05701d1f7405ec2

File: projects/plugins/jetpack/modules/google-fonts.php (hunk 8 above)
Comment (creativecoder) [reply projects/plugins/jetpack/modules/google-fonts.php]: I've added the filter `jetpack_google_fonts_list` (https://github.com/Automattic/jetpack/pull/23045/commits/90aba2484d78e042ffa4b294994facf59eae465f)

File: projects/plugins/jetpack/modules/google-fonts.php (hunk 9 above)
Comment (creativecoder) [reply projects/plugins/jetpack/modules/google-fonts.php]: Changed to `No`. Do you know how we could auto-activate it only on Atomic sites?

Let me know if there are any other module header changes you would recommend. I wasn't sure how to set things like recommendation order.

Comment (jeherve) [reply projects/plugins/jetpack/modules/google-fonts.php]: There are different ways we could go about it, depending on how you'd like to approach the deployment of this feature. Here is what comes to mind:

- If you'd rather not offer any UI, and enable it on all WoA sites right away, we could get rid of this module file in Jetpack, and instead require and instantiate the package from wpcomsh. This is what we do for the post-list package for example; the package isn't used in the Jetpack plugin, but is used in the wpcomsh plugin.
- If you'd like to offer a basic UI (in the legacy full module list) that would eventually turn into a proper setting in the Jetpack dashboard once the feature is deemed ready, you could enable the module on WoA like we do for the masterbar feature in wpcomsh. Check the `feature-plugins` directory for an example.

**Comment by jeherve** [comment]
> Install the Gutenberg plugin with changes in this PR https://github.com/WordPress/gutenberg/pull/37140

Noting that the PR has now been merged.

File: projects/plugins/jetpack/modules/google-fonts.php (hunk 9 above)
Comment (creativecoder) [reply projects/plugins/jetpack/modules/google-fonts.php]: I think it'd be nice to have a minimal UI for the module in Jetpack. I imagine lot's of plugin users will be interested in this feature, outside of WP.com hosted sites.

If you're willing, I'd like to get this basic version merged without a UI, and follow up with auto-activating the module for Atomic sites (wpcomsh) and adding a basic UI (Jetpack).

**Comment by creativecoder** [comment]
Noting that D75501-code (for wpcom testing) is now updated with the latest module and provider changes from this PR.

File: projects/plugins/jetpack/modules/google-fonts.php [hunk 10]
```diff
@@ -0,0 +1,87 @@
+<?php
+/**
+ * Module Name: Google Fonts
+ * Module Description: A selection of Google fonts for block enabled themes.
+ * Sort Order: 1
+ * Recommendation Order: 2
+ * First Introduced: $$next-version$$
+ * Requires Connection: No
```
[the remark this answers is not in the document]
Comment (jeherve) [reply projects/plugins/jetpack/modules/google-fonts.php:8]: That's a good question. I think those privacy implications may be between Google and the site owner, I'm not sure that's something we can cover in our own terms? That may be a better question to ask in +automattlock maybe?

**Review by jeherve — APPROVED** [review]

**Comment by jeherve** [comment]
I've created a draft support doc we can update once we have some more information about the feature and its activation, and I've created #23295 to add a toggle to the Jetpack dashboard to turn the feature on and off.

**Pull request merged**